Makefile

```diff
@@ -1,190 +1,310 @@
+	  make DOCKER_EXTRA_BUILD_TAGS='staging' DOCKER_REGISTRY_HOST=registry.example.com docker-build"
+
+##---- Application -------------------------------------------------------------
+
+## Build the application in the subdirectory (default)
+build:
+	@GOPATH=$(BUILD_DIR_ABS) go install ./cmd/...
+
+## Build the application and install to the system GOPATH
+install:
+	go install ./cmd/...
+
+test_go:
+	@GOPATH=$(BUILD_DIR_ABS) go test ./cmd/...
+
+## Test the application
+test: test_go
+
+# DEPRECATED
+test_all: test
+
+# PostgreSQL should be accessible via postgres:postgres@127.0.0.1:5432/postgres
+# Alternatively override default connection params via TEST_DB_[HOST|PORT|USERNAME|USER_PASSWORD|NAME]
+# Package libpq-dev must be installed for python psycopg2
+## Run extra python test
+test_python:
+	@virtualenv --python=python3 $(BUILD_DIR_ABS)/test_env && \
+		$(BUILD_DIR_ABS)/test_env/bin/pip install -r tests/requirements.txt && \
+		GOPATH=$(BUILD_DIR_ABS) $(BUILD_DIR_ABS)/test_env/bin/python \
+			$(BUILD_DIR_ABS)/src/github.com/cossacklabs/acra/tests/test.py
+
+## Remove build artifacts
+clean:
+	@test -d $(BUILD_DIR_ABS) && chmod -R u+w $(BUILD_DIR_ABS) || true
+	@rm -rf $(BUILD_DIR_ABS)
+
+## Generate keys
 keys: install
 	@chmod +x scripts/generate-keys.sh
 	@scripts/generate-keys.sh
+
+##---- Docker ------------------------------------------------------------------
+
+## Docker : build the image locally
+docker-build:
+	$(call docker_build,acra-build)
+	$(call docker_build,acra-server)
+	$(call docker_build,acra-connector)
+	$(call docker_build,acra-translator)
+	$(call docker_build,acra-keymaker)
+	$(call docker_build,acra-tools)
+	$(call docker_build,acra-webconfig)
+	$(call docker_build,acra-authmanager)
+	$(DOCKER_BIN) image prune --force -a \
+		--filter label=com.cossacklabs.product.name="$(APP_NAME)" \
+		--filter label=com.cossacklabs.docker.container.type="build"
+
+## Docker : tag and push image to remote registry
+docker-push:
+	$(call docker_push,acra-server)
+	$(call docker_push,acra-connector)
+	$(call docker_push,acra-translator)
+	$(call docker_push,acra-keymaker)
+	$(call docker_push,acra-tools)
+	$(call docker_push,acra-webconfig)
+	$(call docker_push,acra-authmanager)
+
+## Docker : remove stopped containers and dangling images
+docker-clean:
+	$(DOCKER_BIN) container prune --force \
+		--filter label=com.cossacklabs.product.name="$(APP_NAME)"
+	$(DOCKER_BIN) image prune --force -a \
+		--filter label=com.cossacklabs.product.name="$(APP_NAME)"
+
+## Docker : alias for 'docker-build' target (DEPRECATED)
+docker: docker-build
+
+##---- Packages ----------------------------------------------------------------
+
+## Package : build deb/rpm depending on the current OS
+pkg: build
+ifeq ($(OS_NAME),unknown)
+	$(error OS is not detected)
+endif
+ifeq ($(OS_CODENAME),unknown)
+	$(error OS codename is not detected)
+endif
+ifeq ($(OS_VERSION_MAJOR),unknown)
+	$(error OS version is not detected)
+endif
```

Comment:
I have a feeling that all variable-setting lines could be placed in the makefile directly instead of being executed only on `make pkg` via `$(eval)`. I don’t think they take that much time but eval makes them harder to read and understand.

There is a point in raising an error when the current OS cannot be detected only on `make pkg`, when we actually care about the OS. This should be done only during `make pkg`, of course.

Also, it would probably look more readable if this were

```make
pkg: build
        tools/make_pkg.sh
```
Now that we add a bulk of code here it’s even more prominent.

Reply:
We discussed the approach previously. While it "not so huge and ugly" it looks preferrable to keep it inside the Makefile (300 lines logically divided into blocks by ≈25 lines already on the verge, but still allow to keep all in one file).

Reply:
> We discussed the approach previously.

“If you have an important point to make, hit it once. Then come back and hit it again. Then hit it a third time.” 😉

(Yes, it’s just annoying to remember to write `\` at the end of lines, properly escape various special symbols, build && giant && chains && of && commands && like && in && Dockerfiles, etc.)